File: Manager.cpp
/*
 *  Manager.cpp
 *  core
 *
 *  Created by Tymon Tobolski on 10-08-23.
 *  Copyright 2010 Politechnika Wrocławska. All rights reserved.
 *
 */

#include <stdio.h>
#include "Manager.h"

#include "Fader.h"

using namespace std;
using namespace TUIO;

Manager::Manager(){
	tuioClient = new TuioClient();
	tuioClient->addTuioListener(this);
	tuioClient->connect();
}


void Manager::setApp(Application *app){
	_app = app;
	_app->start();
}

Application* Manager::app(){
	// TODO: Add throw Exception if _app is null
	return _app;
}

void Manager::display(){
	// display application
	app()->display();
	
	// display objects	
	tuioClient->lockObjectList(); // prevent TUIO thread disaster
	for (list<Object*>::iterator it = objects.begin(); it != objects.end(); ++it){
		(*it)->display();
	}
	
	// display objects connections
	for (list<DirectedConnection*>::iterator it = connections.begin(); it != connections.end(); ++it){
		(*it)->display();
	}
	tuioClient->unlockObjectList();
	

	// display cursors
	tuioClient->lockCursorList(); // prevent TUIO thread disaster
	for (list<Cursor*>::iterator it = cursors.begin(); it != cursors.end(); ++it){
		(*it)->display();
	}
	tuioClient->unlockCursorList();
}



// TUIO callbacks

void Manager::addTuioObject(TuioObject *tobj){
	Object * obj = app()->createObject(tobj);
	obj->move();
	objects.push_back(obj);
}

void Manager::updateTuioObject(TuioObject *tobj){
	tuioClient->lockObjectList(); // prevent TUIO thread disaster

	// update is done automatically via TuioObject pointers in Object instances
	for (list<Object*>::iterator it = objects.begin(); it != objects.end(); ++it){
		if((*it)->checkSource(tobj)) {
			(*it)->move();
		}
	}
	
	
	
	for (list<DirectedConnection*>::iterator it = connections.begin(); it != connections.end(); ++it){
		DirectedConnection * con = *it;
		printf("(%ld, %ld), ", con->from->sid(), con->to->sid());
	}
	
	printf("\n");
	tuioClient->unlockObjectList();
}

void Manager::removeTuioObject(TuioObject *tobj){
	tuioClient->lockObjectList(); // prevent TUIO thread disaster
	
	for (list<Object*>::iterator it = objects.begin(); it != objects.end();){
		Object * obj = *it;
		if(obj->checkSource(tobj)){
			
			for (list<DirectedConnection*>::iterator cit = connections.begin(); cit != connections.end();){
				DirectedConnection * con = *cit;
				if(con->check(obj)){
					cit = connections.erase(cit);
					delete con;
				} else {
					cit++;
				}
			}
			
			
			it = objects.erase(it);
			delete obj;
		} else {
			it++;
		}

	}
	
	tuioClient->unlockObjectList();
}


void Manager::addTuioCursor(TuioCursor *tcur){
	cursors.push_back(app()->createCursor(tcur));
}

void Manager::updateTuioCursor(TuioCursor *tcur){
	// update is done automatically via TuioCursor pointers in Cursor instances
	for (list<Cursor*>::iterator it = cursors.begin(); it != cursors.end(); ++it){
		(*it)->call(E_MOVE);
	}
}

void Manager::removeTuioCursor(TuioCursor *tcur){
	for (list<Cursor*>::iterator it = cursors.begin(); it != cursors.end(); ++it){
		Cursor * cur = *it;
		if(cur->checkSource(tcur)){
			cursors.remove(cur);
			delete cur;
			return;
		}
	}
}


void Manager::refresh(TuioTime ftime){
	//log("Refresh");
}


// tools
list<Object *> Manager::findCloseObjects(Node<> * node, int range, const std::type_info &type){
	list<Object *> close;
	
	for (list<Object*>::iterator it = objects.begin(); it != objects.end(); ++it){
		if(node != (Node<>*)(*it) && typeid(**it) == type && node->distanceTo((Node<>*)*it) <= range){
			close.push_back(*it);
		}
	}
	
	return close;
}

list<Object *> Manager::findCloseObjects(Node<> * node, int range){
	// TODO: Implement this
	list<Object *> close;
	
	for (list<Object*>::iterator it = objects.begin(); it != objects.end(); ++it){
		if(node != (Node<>*)(*it) && node->distanceTo((Node<>*)*it) <= range){
			close.push_back(*it);
		}
	}
	
	return close;
}

list<Cursor *> Manager::findCloseCursors(Node<> * node, int range){
	// TODO: Implement this
	list<Cursor *> close;
	
	for (list<Cursor*>::iterator it = cursors.begin(); it != cursors.end(); ++it){
		if(node != (Node<>*)(*it) && node->distanceTo((Node<>*)*it) <= range){
			close.push_back(*it);
		}
	}
		
	return close;
}

void Manager::addConnection(Object * from, Object * to){
	connections.push_back(new DirectedConnection(from, to));
}

void Manager::removeConnection(Object * from, Object * to){
	for (list<DirectedConnection*>::iterator it = connections.begin(); it != connections.end(); ++it){
		DirectedConnection * con = *it;
		if(con->check(from, to)){
			connections.remove(con);
			delete con;
			return;
		}
	}
}

File: src/main/scala/synergia/platri/Manager.scala
package synergia.platri

import scala.collection.mutable.{ListBuffer, HashMap}
import TUIO._

object Manager extends TuioListener with GFX {
    protected var application: Option[Application] = None

    val tuio = new TuioClient
    tuio.addTuioListener(this)
    tuio.connect

    View // init the object

    val objects = new HashMap[TuioObject, Object]
    val cursors = new HashMap[TuioCursor, Cursor]
    val connections = new ListBuffer[Connection]
    val graphics = new ListBuffer[GFX]
    val topGraphics = new ListBuffer[GFX]

    def run(app: Application){
        application = Some(app)
        app.start
    }

    def stop {
        application.foreach(_.stop)
    }

    def display {
        Calibration.display
        Debug.display
        application.foreach(_.display)
        (graphics ++ connections ++ objects.values ++ cursors.values ++ topGraphics).foreach(_.display)
    }

    def removeConnection(from: Object, to: Object){
        connections.filter(_.check(from, to)).foreach { conn =>
            conn.onRemoved
            connections -= conn
        }
    }

    def addTuioObject(tobj: TuioObject) {
        Debug.info("app: " + application)

        application.foreach { app =>
            val obj = app.createObject(tobj)
            obj.move
            objects(tobj) = obj
        }
    }

    def updateTuioObject(tobj: TuioObject) {
        objects(tobj).move
    }

    def removeTuioObject(tobj: TuioObject) {
        val obj = objects(tobj)
        obj.remove
        connections.filter(_.check(obj)).foreach(connections -= _)
        objects -= tobj
    }

    def addTuioCursor(tcur: TuioCursor) {
        application.foreach { app =>
            cursors(tcur) = app.createCursor(tcur)
        }
    }

    def updateTuioCursor(tcur: TuioCursor) {
        cursors(tcur).move
    }

    def removeTuioCursor(tcur: TuioCursor) {
        cursors -= tcur
    }

    def refresh(ftime: TuioTime) {

    }

    def findCloseObjects(node: Node, range: Int) = objects.values.filter(o => o != node && node.distanceTo(o) <= range)

}

object Main {
    def main(args: Array[String]) {
        Manager.run(apps.boxee.App)
    }
}
